crates/core/asset/src/balance.rs

```diff
@@ -26,18 +26,86 @@ mod imbalance;
 mod iter;
 use commitment::VALUE_BLINDING_GENERATOR;
 use decaf377::{r1cs::ElementVar, Fq, Fr};
-use imbalance::Imbalance;
+use imbalance::{Imbalance, Sign};
 
 use self::commitment::BalanceCommitmentVar;
+use penumbra_proto::{penumbra::core::asset::v1 as pb, DomainType};
 
 /// A `Balance` is a "vector of [`Value`]s", where some values may be required, while others may be
 /// provided. For a transaction to be valid, its balance must be zero.
-#[serde_as]
 #[derive(Clone, Eq, Default, Serialize, Deserialize)]
+#[serde(try_from = "pb::Balance", into = "pb::Balance")]
 pub struct Balance {
-    negated: bool,
-    #[serde_as(as = "Vec<(_, _)>")]
-    balance: BTreeMap<Id, Imbalance<NonZeroU128>>,
+    pub negated: bool,
+    pub balance: BTreeMap<Id, Imbalance<NonZeroU128>>,
+}
+
+impl Balance {
+    fn from_signed_value(negated: bool, value: Value) -> Option<Self> {
+        let non_zero = NonZeroU128::try_from(value.amount.value()).ok()?;
+        Some(Self {
+            negated: false,
+            balance: BTreeMap::from([(
+                value.asset_id,
+                if negated {
+                    Imbalance::Required(non_zero)
+                } else {
+                    Imbalance::Provided(non_zero)
+                },
+            )]),
+        })
+    }
+}
+
+impl DomainType for Balance {
+    type Proto = pb::Balance;
+}
+
+/// Serialization should normalize the `Balance`, where the top-level
+/// negated field is excluded during serialization. Rather, the
+/// sign information is captured in the `SignedValue` pairs.  
+///
+/// Since the underlying BTreeMap can't hold multiple imbalances for
+/// the same asset ID, we implement an ordering-agnostic accumulation
+/// scheme that explicitly combines imbalances.
+impl TryFrom<pb::Balance> for Balance {
+    type Error = anyhow::Error;
+
+    fn try_from(balance: pb::Balance) -> Result<Self, Self::Error> {
+        let mut out = Self::default();
+        for v in balance.values {
+            let value = v.value.ok_or_else(|| anyhow!("missing value"))?;
+            if let Some(b) = Balance::from_signed_value(v.negated, Value::try_from(value)?) {
+                out += b;
+            }
+        }
+        Ok(out)
+    }
+}
+
+impl From<Balance> for pb::Balance {
+    fn from(v: Balance) -> Self {
+        let values = v
+            .balance
+            .into_iter()
+            .map(|(id, imbalance)| {
+                // Decompose imbalance into it sign and magnitude, and convert
+                // magnitude into raw amount and determine negation based on the sign.
+                let (sign, magnitude) = if v.negated { -imbalance } else { imbalance }.into_inner();
```

Comment:
ACK, we apply the negated and collect the resulting sign. That makes sense to me